README.md
# go-sqlite

[![Go Reference](https://pkg.go.dev/badge/github.com/lanz-dev/go-sqlite.svg)](https://pkg.go.dev/github.com/lanz-dev/go-sqlite)
[![Coverage Status](https://coveralls.io/repos/github/lanz-dev/go-sqlite/badge.svg?branch=main)](https://coveralls.io/github/lanz-dev/go-sqlite?branch=main)
[![Github Action](https://github.com/lanz-dev/go-sqlite/actions/workflows/main.yml/badge.svg)](https://github.com/lanz-dev/go-sqlite/actions/workflows/main.yml)

Package sqlite is a collection of tools to work with a SQLite database in go.

# Example

This is a typical usage example.

	package main

	import (
		_ "github.com/mattn/go-sqlite3"
		// or
		_ "modernc.org/sqlite"

		"github.com/lanz-dev/go-sqlite"
	)

	func main() {
		db, err := sqlite.Connect(
			sqlite.WithPath("file:data.db"),
		)
		if err != nil {
			panic(err)
		}
		db.Exec("SELECT 1")
	}


CHANGELOG.md
# Changelog

v0.1.0
- Initial Release


config_test.go
package sqlite

import (
	"errors"
	"testing"
)

func Test_newConfig_Defaults(t *testing.T) {
	t.Parallel()

	config := newConfig()
	if config.LimitConnection != true {
		t.Errorf("expected '%v', got '%v'", true, config.LimitConnection)
	}
	if config.BusyTimeout != 4000 {
		t.Errorf("expected '%d', got '%d'", 4000, config.BusyTimeout)
	}
	if config.ForeignKey != true {
		t.Errorf("expected '%v', got '%v'", true, config.ForeignKey)
	}
	if config.JournalMode != JournalWAL {
		t.Errorf("expected '%s', got '%s'", JournalWAL, config.JournalMode)
	}
	if config.JournalSizeLimit != 100000000 {
		t.Errorf("expected '%d', got '%d'", 100000000, config.JournalSizeLimit)
	}
	if config.SyncMode != SyncNormal {
		t.Errorf("expected '%s', got '%s'", SyncNormal, config.SyncMode)
	}
}

func Test_buildConfig_DefaultPath(t *testing.T) {
	t.Parallel()

	config, err := buildConfig(
		WithDriver(DriverModernc),
	)
	if err != nil {
		t.Fatalf("did not expect error '%s'", err)
	}
	expected := ":memory"
	if config.Path != expected {
		t.Fatalf("expected '%s', got '%s'", expected, config.Path)
	}
}

func Test_buildConfig_InvalidPath(t *testing.T) {
	t.Parallel()

	_, err := buildConfig(
		WithPath("invalid"),
		WithDriver(DriverModernc),
	)
	if !errors.Is(err, ErrInvalidPath) {
		t.Fatalf("expected to receive error '%s', got '%s'", ErrInvalidPath, err)
	}
}

func Test_buildConfig_DriverNameModernc(t *testing.T) {
	t.Parallel()

	config, err := buildConfig(
		WithDriver(DriverModernc),
	)
	if err != nil {
		t.Fatalf("did not expect error '%s'", err)
	}
	expected := DriverNameModernc
	if config.DriverName != expected {
		t.Fatalf("expected '%s', got '%s'", expected, config.DriverName)
	}
}

func Test_buildConfig_DriverNameMattn(t *testing.T) {
	t.Parallel()

	config, err := buildConfig(
		WithDriver(DriverMattn),
	)
	if err != nil {
		t.Fatalf("did not expect error '%s'", err)
	}
	expected := DriverNameMattn
	if config.DriverName != expected {
		t.Fatalf("expected '%s', got '%s'", expected, config.DriverName)
	}
}

func Test_buildConfig_DSNModernc(t *testing.T) {
	t.Parallel()

	config, err := buildConfig(
		WithDriver(DriverModernc),
	)
	if err != nil {
		t.Fatalf("did not expect error '%s'", err)
	}
	expected := ":memory?_pragma=busy_timeout(4000)&_pragma=foreign_keys(1)&_pragma=journal_mode(WAL)&_pragma=synchronous(NORMAL)"
	if config.DSN != expected {
		t.Fatalf("expected '%s', got '%s'", expected, config.DSN)
	}
}

func Test_buildConfig_DSNMattn(t *testing.T) {
	t.Parallel()

	config, err := buildConfig(
		WithDriver(DriverMattn),
	)
	if err != nil {
		t.Fatalf("did not expect error '%s'", err)
	}
	expected := ":memory?_timeout=4000&_fk=true&_journal=WAL&_sync=1"
	if config.DSN != expected {
		t.Fatalf("expected '%s', got '%s'", expected, config.DSN)
	}
}

func TestAutoVacuumMode_Int(t *testing.T) {
	tests := []struct {
		name string
		s    AutoVacuumMode
		want int
	}{
		{"AutoVaccumNone", AutoVacuumNone, 0},
		{"AutoVacuumFull", AutoVacuumFull, 1},
		{"AutoVacuumIncremental", AutoVacuumIncremental, 2},
		{"AutoVacuumInvalid", AutoVacuumMode("invalid"), -99},
	}
	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			if got := tc.s.Int(); got != tc.want {
				t.Errorf("Int() = %v, want %v", got, tc.want)
			}
		})
	}
}

func TestSyncMode_Int(t *testing.T) {
	tests := []struct {
		name string
		s    SyncMode
		want int
	}{
		{"SyncOff", SyncOff, 0},
		{"SyncNormal", SyncNormal, 1},
		{"SyncFull", SyncFull, 2},
		{"SyncExtra", SyncExtra, 3},
		{"SyncInvalid", SyncMode("invalid"), -99},
	}
	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			if got := tc.s.Int(); got != tc.want {
				t.Errorf("Int() = %v, want %v", got, tc.want)
			}
		})
	}
}

func Test_validatePath(t *testing.T) {
	tests := []struct {
		path    string
		wantErr error
	}{
		{"invalid:/path", ErrInvalidPath},
		{"invalid", ErrInvalidPath},
		{"file", ErrInvalidPath},
		{"file:", ErrInvalidPath},
		{"file:/", ErrInvalidPath},
		{"file:/fsafd", ErrInvalidPath},
		{"file:/data", ErrInvalidPath},
		{"file:/data.", ErrInvalidPath},
		{"file:/data.db", nil},
		{"file:data.db", nil},
		{"file:../../../../../../data.db", nil},
		{":memory", nil},
	}
	for _, tc := range tests {
		tc := tc

		t.Run("With path '"+tc.path+"'", func(t *testing.T) {
			t.Parallel()

			err := validatePath(tc.path)
			if err != nil && tc.wantErr == nil {
				t.Fatalf("did not expect error '%s'", err)
			}
			if tc.wantErr != nil && !errors.Is(err, tc.wantErr) {
				t.Fatalf("expect error to be '%s', got '%s'", tc.wantErr, err)
			}
			if tc.wantErr == nil && err != nil {
				t.Fatalf("expected '%s' to be valid", tc.path)
			}
		})
	}
}


sqlite_test.go
package sqlite_test

import (
	"database/sql"
	"errors"
	"testing"

	"github.com/DATA-DOG/go-sqlmock"

	"github.com/lanz-dev/go-sqlite"
)

func buildMockDB(t *testing.T) (*sql.DB, sqlmock.Sqlmock) {
	db, mock, err := sqlmock.New()
	if err != nil {
		t.Fatalf("an error '%s' was not expected when opening a stub database connection", err)
	}
	return db, mock
}

func TestOptimize(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	mock.ExpectExec("PRAGMA optimize;").
		WillReturnResult(sqlmock.NewResult(1, 1))

	if err := sqlite.Optimize(db); err != nil {
		t.Fatalf("did not expected error '%s'", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}

func TestOptimize_WithError(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	errUnitTest := errors.New("unittest")

	mock.ExpectExec("PRAGMA optimize;").
		WillReturnError(errUnitTest)

	err := sqlite.Optimize(db)
	if !errors.Is(err, errUnitTest) {
		t.Fatalf("expect error to be '%s', got '%s'", errUnitTest, err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}

func TestShutdown(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	mock.ExpectExec("PRAGMA optimize;").
		WillReturnResult(sqlmock.NewResult(1, 1))
	mock.ExpectClose()

	if err := sqlite.Shutdown(db); err != nil {
		t.Fatalf("did not expected error '%s'", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}

func TestShutdown_WithErrorInOptimize(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	errUnitTest := errors.New("unittest")

	mock.ExpectExec("PRAGMA optimize;").
		WillReturnError(errUnitTest)

	err := sqlite.Shutdown(db)
	if !errors.Is(err, errUnitTest) {
		t.Fatalf("expect error to be '%s', got '%s'", errUnitTest, err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}

func TestShutdown_WithErrorOnClose(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	errUnitTest := errors.New("unittest")

	mock.ExpectExec("PRAGMA optimize;").
		WillReturnResult(sqlmock.NewResult(1, 1))
	mock.ExpectClose().WillReturnError(errUnitTest)

	err := sqlite.Shutdown(db)
	if !errors.Is(err, errUnitTest) {
		t.Fatalf("expect error to be '%s', got '%s'", errUnitTest, err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}

func TestVacuum(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	mock.ExpectExec("VACUUM;").
		WillReturnResult(sqlmock.NewResult(1, 1))

	if err := sqlite.Vacuum(db); err != nil {
		t.Fatalf("did not expected error '%s'", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}

func TestVacuum_WithError(t *testing.T) {
	db, mock := buildMockDB(t)
	defer db.Close()

	errUnitTest := errors.New("unittest")

	mock.ExpectExec("VACUUM;").
		WillReturnError(errUnitTest)

	err := sqlite.Vacuum(db)
	if !errors.Is(err, errUnitTest) {
		t.Fatalf("expect error to be '%s', got '%s'", errUnitTest, err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Errorf("there were unfulfilled expectations: %s", err)
	}
}


options_test.go
package sqlite

import (
	"testing"
)

func optionRunner(config *Config, opts ...Option) {
	for _, opt := range opts {
		opt(config)
	}
}

func TestWithAutoVacuumMode(t *testing.T) {
	t.Parallel()

	expected := AutoVacuumFull

	config := newConfig()
	optionRunner(
		config,
		WithAutoVacuumMode(expected),
	)

	got := config.AutoVacuumMode
	if got != expected {
		t.Errorf("expected '%s', got '%s'", expected, got)
	}
}

func TestWithBusyTimeout(t *testing.T) {
	t.Parallel()

	expected := 100

	config := newConfig()
	optionRunner(
		config,
		WithBusyTimeout(expected),
	)

	got := config.BusyTimeout
	if got != expected {
		t.Errorf("expected '%d', got '%d'", expected, got)
	}
}

func TestWithCaseSensitiveLike(t *testing.T) {
	t.Parallel()

	config := newConfig()
	optionRunner(
		config,
		WithCaseSensitiveLike(true),
	)

	got := config.CaseSensitiveLike
	if !got {
		t.Errorf("expected '%v', got '%v'", true, got)
	}
}

func TestWithDeferredForeignKeys(t *testing.T) {
	t.Parallel()

	config := newConfig()
	optionRunner(
		config,
		WithDeferredForeignKeys(true),
	)

	got := config.DeferForeignKeys
	if !got {
		t.Errorf("expected '%v', got '%v'", true, got)
	}
}

func TestWithDisabledLimits(t *testing.T) {
	t.Parallel()

	config := newConfig()
	optionRunner(
		config,
		WithDisabledLimits(),
	)

	got := config.LimitConnection
	if got {
		t.Errorf("expected '%v', got '%v'", false, got)
	}
}

func TestWithDriver(t *testing.T) {
	t.Parallel()

	expected := DriverModernc

	config := newConfig()
	optionRunner(
		config,
		WithDriver(expected),
	)

	got := config.Driver
	if got != expected {
		t.Errorf("expected '%s', got '%s'", expected, got)
	}
}

func TestWithDriverName(t *testing.T) {
	t.Parallel()

	expected := DriverNameModernc

	config := newConfig()
	optionRunner(
		config,
		WithDriverName(expected),
	)

	got := config.DriverName
	if got != expected {
		t.Errorf("expected '%s', got '%s'", expected, got)
	}
}

func TestWithForeignKeySupport(t *testing.T) {
	t.Parallel()

	config := newConfig()
	optionRunner(
		config,
		WithForeignKeySupport(false),
	)

	got := config.ForeignKey
	if got {
		t.Errorf("expected '%v', got '%v'", false, got)
	}
}

func TestWithJournalMode(t *testing.T) {
	t.Parallel()

	expected := JournalPersist

	config := newConfig()
	optionRunner(
		config,
		WithJournalMode(expected),
	)

	got := config.JournalMode
	if got != expected {
		t.Errorf("expected '%s', got '%s'", expected, got)
	}
}

func TestWithJournalSizeLimit(t *testing.T) {
	t.Parallel()

	expected := 100

	config := newConfig()
	optionRunner(
		config,
		WithJournalSizeLimit(expected),
	)

	got := config.JournalSizeLimit
	if got != expected {
		t.Errorf("expected '%d', got '%d'", expected, got)
	}
}

func TestWithPath(t *testing.T) {
	t.Parallel()

	expected := "file:/data.db"

	config := newConfig()
	optionRunner(
		config,
		WithPath(expected),
	)

	got := config.Path
	if got != expected {
		t.Errorf("expected '%s', got '%s'", expected, got)
	}
}

func TestWithSyncMode(t *testing.T) {
	t.Parallel()

	expected := SyncExtra

	config := newConfig()
	optionRunner(
		config,
		WithSyncMode(expected),
	)

	got := config.SyncMode
	if got != expected {
		t.Errorf("expected '%s', got '%s'", expected, got)
	}
}


config.go
package sqlite

import (
	"fmt"
	"regexp"
)

// AutoVacuumMode for the SQLite connection.
//
// See https://www.sqlite.org/pragma.html#pragma_auto_vacuum.
type AutoVacuumMode string

// Int will return the int value for "github.com/mattn/go-sqlite3".
func (s *AutoVacuumMode) Int() int {
	switch *s {
	case AutoVacuumNone:
		return 0
	case AutoVacuumFull:
		return 1
	case AutoVacuumIncremental:
		return 2
	}
	return -99
}

// Driver used to connect to SQLite.
type Driver string

// JournalMode for the SQLite Connection.
//
// See https://www.sqlite.org/pragma.html#pragma_journal_mode.
type JournalMode string

// SyncMode for the SQLite Connection.
//
// See https://www.sqlite.org/pragma.html#pragma_synchronous.
type SyncMode string

// Int will return the int value for "github.com/mattn/go-sqlite3".
func (s *SyncMode) Int() int {
	switch *s {
	case SyncOff:
		return 0
	case SyncNormal:
		return 1
	case SyncFull:
		return 2
	case SyncExtra:
		return 3
	}
	return -99
}

// The different available AutoVaccum modes for SQLite.
//
// See https://www.sqlite.org/pragma.html#pragma_auto_vacuum.
const (
	AutoVacuumDefault     AutoVacuumMode = ""
	AutoVacuumNone        AutoVacuumMode = "NONE"
	AutoVacuumFull        AutoVacuumMode = "FULL"
	AutoVacuumIncremental AutoVacuumMode = "INCREMENTAL"
)

// Based on the Driver the connection DSN to SQLite is build.
const (
	DriverMattn   Driver = "github.com/mattn/go-sqlite3"
	DriverModernc Driver = "modernc.org/sqlite"
)

// These are the default [sql.Open] driverNames for the SQLite drivers.
const (
	DriverNameMattn   = "sqlite3"
	DriverNameModernc = "sqlite"
)

// The different available Journal modes for SQLite.
//
// See https://www.sqlite.org/pragma.html#pragma_journal_mode.
const (
	JournalDefault  JournalMode = ""
	JournalDelete   JournalMode = "DELETE"
	JournalTruncate JournalMode = "TRUNCATE"
	JournalPersist  JournalMode = "PERSIST"
	JournalMemory   JournalMode = "MEMORY"
	JournalWAL      JournalMode = "WAL"
	JournalOff      JournalMode = "OFF"
)

// The different available Sync modes for SQLite.
//
// See https://www.sqlite.org/pragma.html#pragma_synchronous.
const (
	SyncDefault SyncMode = ""
	SyncOff     SyncMode = "OFF"
	SyncNormal  SyncMode = "NORMAL"
	SyncFull    SyncMode = "FULL"
	SyncExtra   SyncMode = "EXTRA"
)

var regexPath = regexp.MustCompile(`^file:.+\..+$`)

// Config for the SQLite connection.
type Config struct {
	DSN             string // DSN string for [sql.Open]
	Driver          Driver // [sqlite.DriverMattn] or [sqlite.DriverModernc]
	DriverName      string // DriverName used in [sql.Open]
	Path            string // Path to the SQLite database
	LimitConnection bool   // Should we set the default limits?

	AutoVacuumMode    AutoVacuumMode // https://www.sqlite.org/pragma.html#pragma_auto_vacuum
	BusyTimeout       int            // https://www.sqlite.org/pragma.html#pragma_busy_timeout
	CaseSensitiveLike bool           // https://www.sqlite.org/pragma.html#pragma_case_sensitive_like
	DeferForeignKeys  bool           // https://www.sqlite.org/pragma.html#pragma_defer_foreign_keys
	ForeignKey        bool           // https://www.sqlite.org/pragma.html#pragma_foreign_keys
	JournalMode       JournalMode    // https://www.sqlite.org/pragma.html#pragma_journal_mode
	JournalSizeLimit  int            // https://www.sqlite.org/pragma.html#pragma_journal_size_limit
	SyncMode          SyncMode       // https://www.sqlite.org/pragma.html#pragma_synchronous
}

func newConfig() *Config {
	return &Config{
		LimitConnection: true,

		BusyTimeout:      4000,
		ForeignKey:       true,
		JournalMode:      JournalWAL,
		JournalSizeLimit: 100000000,
		SyncMode:         SyncNormal,
	}
}

func validatePath(path string) error {
	if path == ":memory" {
		return nil
	}
	if regexPath.MatchString(path) {
		return nil
	}
	return fmt.Errorf("given '%s', %w", path, ErrInvalidPath)
}

func buildConfig(opts ...Option) (*Config, error) {
	config := newConfig()
	for _, opt := range opts {
		opt(config)
	}

	if config.Path == "" {
		config.Path = ":memory"
	}
	if err := validatePath(config.Path); err != nil {
		return nil, err
	}

	if config.Driver == "" {
		var err error
		if config.Driver, err = detectDriver(); err != nil {
			return nil, err
		}
	}

	if config.DriverName == "" {
		if config.Driver == DriverModernc {
			config.DriverName = DriverNameModernc
		} else if config.Driver == DriverMattn {
			config.DriverName = DriverNameMattn
		}
	}

	if config.Driver == DriverModernc {
		config.DSN = buildModerncDSN(config)
	} else if config.Driver == DriverMattn {
		config.DSN = buildMattnDSN(config)
	}

	return config, nil
}


options.go
package sqlite

// Option is a func to set configuration options for SQLite.
type Option func(c *Config)

// WithAutoVacuumMode will set the auto vacuum mode.
//
// Setting the value [sqlite.AutoVacuumDefault] will not set the pragma at all and uses the driver default behaviour.
//
// See https://www.sqlite.org/pragma.html#pragma_auto_vacuum.
func WithAutoVacuumMode(mode AutoVacuumMode) Option {
	return func(c *Config) {
		c.AutoVacuumMode = mode
	}
}

// WithBusyTimeout will set the busy timeout.
//
// Setting a value of 0 will not set the pragma at all and uses the driver default behaviour.
//
// See https://www.sqlite.org/pragma.html#pragma_busy_timeout.
func WithBusyTimeout(timout int) Option {
	return func(c *Config) {
		c.BusyTimeout = timout
	}
}

// WithCaseSensitiveLike will enable or disable the case-sensitive like.
//
// See https://www.sqlite.org/pragma.html#pragma_case_sensitive_like.
func WithCaseSensitiveLike(enabled bool) Option {
	return func(c *Config) {
		c.CaseSensitiveLike = enabled
	}
}

// WithDeferredForeignKeys will enable or disable deferred foreign keys.
//
// See https://www.sqlite.org/pragma.html#pragma_defer_foreign_keys.
func WithDeferredForeignKeys(enabled bool) Option {
	return func(c *Config) {
		c.DeferForeignKeys = enabled
	}
}

// WithDisabledLimits will disable the default limits for the connection.
//
// These are the default limits:
//   - db.SetMaxIdleConns(1)
//   - db.SetMaxOpenConns(1)
//   - db.SetConnMaxLifetime(0)
//   - db.SetConnMaxIdleTime(0)
func WithDisabledLimits() Option {
	return func(c *Config) {
		c.LimitConnection = false
	}
}

// WithDriver will enforce the usage of a specific [driver.Driver].
//
// Available options are:
//   - [sqlite.DriverMattn] for "github.com/mattn/go-sqlite3"
//   - [sqlite.DriverModernc] for "modernc.org/sqlite"
func WithDriver(driver Driver) Option {
	return func(c *Config) {
		c.Driver = driver
	}
}

// WithDriverName will set the name used for [sql.Open].
//
// Normally you just need this, if you modified the sql [driver.Driver]
// This happens normally if you want to add tracing, profiling, etc.
func WithDriverName(name string) Option {
	return func(c *Config) {
		c.DriverName = name
	}
}

// WithForeignKeySupport will enable or disable the foreign key support.
//
// See https://www.sqlite.org/pragma.html#pragma_foreign_keys.
func WithForeignKeySupport(enabled bool) Option {
	return func(c *Config) {
		c.ForeignKey = enabled
	}
}

// WithJournalMode will set the journal mode for the connection.
//
// Setting the value [sqlite.JournalDefault] will not set the pragma at all and uses the driver default behaviour.
//
// See https://www.sqlite.org/pragma.html#pragma_journal_mode.
func WithJournalMode(mode JournalMode) Option {
	return func(c *Config) {
		c.JournalMode = mode
	}
}

// WithJournalSizeLimit will set the journal size limit.
//
// Setting a value of 0 will not set the pragma at all and uses the driver default behaviour.
//
// See https://www.sqlite.org/pragma.html#pragma_journal_size_limit.
func WithJournalSizeLimit(limit int) Option {
	return func(c *Config) {
		c.JournalSizeLimit = limit
	}
}

// WithPath will set the db path for sqlite
//
// dbPath should be in format "file:your/path/to/data.db" or ":memory" for an in-memory sqlite connection.
// The format will be checked per regex `^file\:.+\..+$` on [sqlite.Connect].
func WithPath(dbPath string) Option {
	return func(c *Config) {
		c.Path = dbPath
	}
}

// WithSyncMode will set the sync mode for the connection.
//
// Setting the value [sqlite.SyncDefault] will not set the pragma at all and uses the driver default behaviour.
//
// See https://www.sqlite.org/pragma.html#pragma_synchronous.
func WithSyncMode(sync SyncMode) Option {
	return func(c *Config) {
		c.SyncMode = sync
	}
}


sqlite_connect.go
package sqlite

import (
	"database/sql"
	"fmt"
	_ "unsafe" // For go:linkname
)

func buildDSN(path string, params []string) string {
	dsn := path
	if len(params) == 0 {
		return dsn
	}

	dsn += "?" + params[0]
	for _, p := range params[1:] {
		dsn += "&" + p
	}

	return dsn
}

// buildMattnDSN works with "github.com/mattn/go-sqlite3"
func buildMattnDSN(config *Config) string {
	var params []string

	if config.AutoVacuumMode != AutoVacuumDefault {
		mode := config.AutoVacuumMode.Int()
		if mode != -99 {
			params = append(params, fmt.Sprintf("_auto_vacuum=%d", mode))
		}
	}
	if config.BusyTimeout > 0 {
		params = append(params, fmt.Sprintf("_timeout=%d", config.BusyTimeout))
	}
	if config.CaseSensitiveLike {
		params = append(params, "_case_sensitive_like=true")
	}
	if config.ForeignKey {
		params = append(params, "_fk=true")
		if config.DeferForeignKeys {
			params = append(params, "defer_foreign_keys=true")
		}
	}
	if config.JournalMode != JournalDefault {
		params = append(params, fmt.Sprintf("_journal=%s", config.JournalMode))
	}
	if config.SyncMode != SyncDefault {
		mode := config.SyncMode.Int()
		if mode != -99 {
			params = append(params, fmt.Sprintf("_sync=%d", mode))
		}
	}

	return buildDSN(config.Path, params)
}

// buildModerncDSN works with "modernc.org/sqlite"
func buildModerncDSN(config *Config) string {
	var params []string

	if config.AutoVacuumMode != AutoVacuumDefault {
		params = append(params, fmt.Sprintf("_pragma=auto_vacuum(%s)", config.AutoVacuumMode))
	}
	if config.BusyTimeout > 0 {
		params = append(params, fmt.Sprintf("_pragma=busy_timeout(%d)", config.BusyTimeout))
	}
	if config.CaseSensitiveLike {
		params = append(params, "_pragma=case_sensitive_like(1)")
	}
	if config.ForeignKey {
		params = append(params, "_pragma=foreign_keys(1)")
		if config.DeferForeignKeys {
			params = append(params, "_pragma=defer_foreign_keys(1)")
		}
	}
	if config.JournalMode != JournalDefault {
		params = append(params, fmt.Sprintf("_pragma=journal_mode(%s)", config.JournalMode))
	}
	if config.SyncMode != SyncDefault {
		params = append(params, fmt.Sprintf("_pragma=synchronous(%s)", config.SyncMode))
	}

	return buildDSN(config.Path, params)
}

func detectDriver() (Driver, error) {
	for _, d := range sql.Drivers() {
		if d == DriverNameModernc {
			return DriverModernc, nil
		}
		if d == DriverNameMattn {
			return DriverMattn, nil
		}
	}
	return "", ErrCantDetectDriver
}

type sqlOpenFunc func(driverName, dataSourceName string) (*sql.DB, error)

func connect(openFunc sqlOpenFunc, opts ...Option) (*sql.DB, error) {
	config, err := buildConfig(opts...)
	if err != nil {
		return nil, err
	}

	db, err := openFunc(config.DriverName, config.DSN)
	if err != nil {
		return nil, err
	}
	if err := db.Ping(); err != nil {
		return nil, err
	}

	if config.LimitConnection {
		db.SetMaxIdleConns(1)
		db.SetMaxOpenConns(1)
		db.SetConnMaxLifetime(0)
		db.SetConnMaxIdleTime(0)
	}

	if config.JournalSizeLimit > 0 {
		_, err = db.Exec(fmt.Sprintf("PRAGMA journal_size_limit = %d;", config.JournalSizeLimit))
		if err != nil {
			return nil, err
		}
	}

	return db, err
}


sqlite.go
package sqlite

import (
	"context"
	"database/sql"
	"errors"
)

// ErrCantDetectDriver will be returned if no specific [sqlite.Driver] was set per [sqlite.WithDriver] and
// a registered driver could not be found automatically.
//
// Did you import your driver?
//
//	  import (
//		  _ "github.com/mattn/go-sqlite3"
//		  // or
//		  _ "modernc.org/sqlite"
//	  )
var ErrCantDetectDriver = errors.New("cannot detect the correct driver")

// ErrInvalidPath will be returned if the provided path is invalid.
var ErrInvalidPath = errors.New("invalid path provided")

var openFunc sqlOpenFunc = sql.Open

// Connect will connect to a SQLite database with some typical performance settings and foreign key support.
//
// You should call Shutdown / ShutdownContext on defer or within a shutdown hook.
// You should call after some hours Optimize / OptimizeContext on a regular basis.
//
// If no [sqlite.Driver] is manually set per [sqlite.WithDriver], [sqlite.Connect] tries to detect a [sqlite.Driver].
// The detected [sqlite.Driver] is based on the registered sql drivers:
//   - "sqlite" for "modernc.org/sqlite"
//   - "sqlite3" for "github.com/mattn/go-sqlite3"
//
// These are the default Settings]:
//   - Path is ":memory" for an in-memory sqlite connection
//   - JournalMode WAL
//   - SyncMode NORMAL
//   - Sets on sql.DB the connection limit to 1 and lifetime to 0
//   - Foreign key support enabled
//   - BusyTimeout set to 4000
//   - JournalSizeLimit set to 100000000
func Connect(opts ...Option) (*sql.DB, error) {
	return connect(openFunc, opts...)
}

// Optimize will run `PRAGMA optimize` which should be run on connection close and every few hours.
//
// See https://www.sqlite.org/pragma.html#pragma_optimize
func Optimize(db *sql.DB) error {
	return OptimizeContext(context.Background(), db)
}

// OptimizeContext will run `PRAGMA optimize` which should be run on connection close and every few hours.
//
// See https://www.sqlite.org/pragma.html#pragma_optimize
func OptimizeContext(ctx context.Context, db *sql.DB) error {
	if _, err := db.ExecContext(ctx, "PRAGMA optimize;"); err != nil {
		return err
	}
	return nil
}

// Shutdown should be called before the application exits.
func Shutdown(db *sql.DB) error {
	return ShutdownContext(context.Background(), db)
}

// ShutdownContext should be called before the application exits.
func ShutdownContext(ctx context.Context, db *sql.DB) error {
	if err := OptimizeContext(ctx, db); err != nil {
		return err
	}
	if err := db.Close(); err != nil {
		return err
	}
	return nil
}

// Vacuum will call the VACUUM statement.
//
// This is usefully if you don't have auto-vacuum enabled and want to shrink the sqlite db file.
//
// See https://www.sqlite.org/lang_vacuum.html.
func Vacuum(db *sql.DB) error {
	return VacuumContext(context.Background(), db)
}

// VacuumContext will call the VACUUM statement.
//
// This is usefully if you don't have auto-vacuum enabled and want to shrink the sqlite db file.
//
// See https://www.sqlite.org/lang_vacuum.html.
func VacuumContext(ctx context.Context, db *sql.DB) error {
	if _, err := db.ExecContext(ctx, "VACUUM;"); err != nil {
		return err
	}
	return nil
}
